src/entities/nestedExit/index.ts

```diff
@@ -0,0 +1,147 @@
+            input,
+            nestedPoolState,
+            isProportional,
+        );
+
+        const encodedCalls = encodeCalls(callsAttributes, isProportional);
+
+        const { peekCalls, tokensOut } = getPeekCalls(
+            callsAttributes,
+            isProportional,
+        );
+
+        // append peek calls to get amountsOut
+        encodedCalls.push(...peekCalls);
+
+        const encodedMulticall = encodeFunctionData({
+            abi: balancerRelayerAbi,
+            functionName: 'vaultActionsQueryMulticall',
+            args: [encodedCalls],
+        });
+
+        const peekedValues = await doQueryNestedExit(
+            input.chainId,
+            input.rpcUrl,
+            input.accountAddress,
+            encodedMulticall,
+            tokensOut.length,
+        );
+
+        console.log('peekedValues ', peekedValues);
+
+        const amountsOut = tokensOut.map((tokenOut, i) =>
+            TokenAmount.fromRawAmount(tokenOut, peekedValues[i]),
+        );
+
+        return { callsAttributes, bptAmountIn, amountsOut, isProportional };
+    }
+
+    buildCall(input: NestedExitCallInput): {
+        call: Hex;
+        to: Address;
+        minAmountsOut: TokenAmount[];
+    } {
+        // apply slippage to amountsOut
+        const minAmountsOut = input.amountsOut.map((amountOut) =>
+            TokenAmount.fromRawAmount(
+                amountOut.token,
+                input.slippage.removeFrom(amountOut.amount),
+            ),
+        );
+
+        input.callsAttributes.forEach((call) => {
+            // update relevant calls with minAmountOut limits in place
+            minAmountsOut.forEach((minAmountOut, j) => {
+                const minAmountOutIndex = call.sortedTokens.findIndex((t) =>
+                    t.isSameAddress(minAmountOut.token.address),
+                );
+                if (minAmountOutIndex !== -1) {
+                    call.minAmountsOut[minAmountOutIndex] =
+                        minAmountsOut[j].amount;
+                }
+            });
+        });
+
+        const encodedCalls = encodeCalls(
+            input.callsAttributes,
+            input.isProportional,
+        );
+
+        // prepend relayer approval if provided
+        if (input.relayerApprovalSignature !== undefined) {
+            encodedCalls.unshift(
+                Relayer.encodeSetRelayerApproval(
+                    BALANCER_RELAYER[input.callsAttributes[0].chainId],
+                    true,
+                    input.relayerApprovalSignature,
+                ),
+            );
+        }
+
+        const call = encodeFunctionData({
+            abi: balancerRelayerAbi,
+            functionName: 'multicall',
+            args: [encodedCalls],
+        });
+
+        return {
+            call,
+            to: BALANCER_RELAYER[input.callsAttributes[0].chainId],
+            minAmountsOut,
+        };
+    }
+}
+
+const validateInputs = (
```

Comment:
What do you think about refactoring this a bit to be super clear? Could have two further functions: `validateProportional` and `validateSingleToken`.